support-lib/cpp/SharedFuture.hpp

```diff
@@ -0,0 +1,138 @@
+/**
+ * Copyright 2021 Snap, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+ 
+ #pragma once
+
+#include "Future.hpp"
+
+#include <memory>
+#include <optional>
+#include <type_traits>
+#include <variant>
+#include <vector>
+
+namespace djinni {
+
+// SharedFuture is a wrapper around djinni::Future to allow multiple consumers (i.e. like std::shared_future)
+// The API is designed to be similar to djinni::Future.
+template<typename T>
+class SharedFuture {
+public:
+    // Create SharedFuture from Future. Runtime error if the future is already consumed.
+    explicit SharedFuture(Future<T>&& future);
+
+    // Transform into Future<T>.
+    Future<T> toFuture() const {
+        if (await_ready()) {
+            co_return await_resume(); // return stored value directly
```

Comment:
the whole file needs to be guarded with `#if defined(DJINNI_FUTURE_HAS_COROUTINE_SUPPORT)`